File: src/mtglab/decks/analyze.py
"""Deterministic deck analysis.

Everything here is a pure function of a deck plus a card pool. No DuckDB, no
network, no judgement calls that vary between runs -- the same deck always
produces the same report, and the API layer is a thin wrapper over these.

That is the point. Curve, pip demand and category coverage were previously
worked out by hand in conversation, which meant nobody running the app could
reproduce them and nothing tested them. Logic that lives here gets tests.

The pool is passed in as a `{name: CardRecord}` dict, matching
`decks/validate.py`, so these test without a database.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from mtglab.decks.model import CATEGORIES, Deck
from mtglab.mana import COLORS, parse_mana_cost

if TYPE_CHECKING:
    from mtglab.cards.db import CardRecord

# Rough coverage targets by bracket. These are conventional Commander
# deckbuilding guidance, not computed truth, and are reported as guidance
# rather than enforced -- `validate.py` is the gate, this is the advice.
CATEGORY_TARGETS: dict[str, tuple[int, int]] = {
    "land": (33, 38),
    "ramp": (8, 12),
    "card-advantage": (8, 12),
    "interaction": (6, 12),
    "tutor": (0, 10),
}

#: How many cards from the official Game Changers list each bracket permits.
#: `None` means no limit.
#:
#: Unlike `CATEGORY_TARGETS` above, which is conventional guidance, this is the
#: published rule for the bracket a deck declares — which is why it can be
#: *counted* rather than estimated. Scryfall carries the flag per card, so the
#: count is a card pool fact.
#:
#: It is still reported rather than enforced. `validate.py` is the gate and
#: this module is the advice, and the same reasoning that keeps a banned card a
#: matter for the user to resolve applies here: which Game Changer to cut, or
#: whether to move the deck up a bracket, is a decision rather than a fix.
GAME_CHANGER_LIMITS: dict[int, int | None] = {
    1: 0, 2: 0, 3: 3, 4: None, 5: None,
}


def game_changers(deck: Deck, cards: dict[str, CardRecord] | None = None) -> dict[str, Any]:
    """Which Game Changers the deck runs, and what its bracket allows.

    Returns `allowed: None` for an unlimited bracket and `verdict: "unknown"`
    when the deck declares no bracket or the pool is missing — an absent
    count is not a count of zero, and a deck that reports "0 Game Changers"
    because nobody looked is the quiet wrong answer this project keeps finding.
    """
    cards = cards or {}
    names = [c.name for c in deck.cards] + list(deck.commander)
    if deck.companion:
        names.append(deck.companion)

    found = [n for n in names
             if getattr(cards.get(n), "game_changer", False)]
    allowed = GAME_CHANGER_LIMITS.get(deck.bracket or 0)

    if not cards or deck.bracket is None:
        verdict = "unknown"
    elif allowed is None or len(found) <= allowed:
        verdict = "ok"
    else:
        verdict = "over"

    return {
        "cards": sorted(found),
        "count": len(found),
        "allowed": allowed,
        "bracket": deck.bracket,
        "verdict": verdict,
    }


@dataclass
class CurveBucket:
    mv: int
    count: int
    names: list[str] = field(default_factory=list)


@dataclass
class ColorNeed:
    """Demand for one color, and the supply available to meet it."""

    color: str
    pips: int          # colored pips this deck's spells require, weighted by qty
    sources: int       # permanents/lands that can produce this color
    cards: int         # cards requiring at least one pip of this color

    @property
    def sources_per_pip(self) -> float:
        return round(self.sources / self.pips, 2) if self.pips else 0.0


def _is_land(rec: Any) -> bool:
    return bool(rec) and "Land" in (getattr(rec, "type_line", "") or "")


def curve(deck: Deck, cards: dict[str, CardRecord] | None = None) -> dict[str, Any]:
    """MV histogram over the nonland 99, plus the average nonland MV.

    Lands are excluded: a land has no mana value worth curving, and including
    them drags the average toward zero and makes decks incomparable.
    """
    buckets: dict[int, CurveBucket] = {}
    total_mv = 0
    counted = 0

    for entry in deck.cards:
        rec = (cards or {}).get(entry.name)
        if entry.category == "land" or _is_land(rec):
            continue
        cost = entry.mana_cost or (getattr(rec, "mana_cost", None) if rec else None)
        mv = parse_mana_cost(cost).mana_value
        bucket = buckets.setdefault(mv, CurveBucket(mv=mv, count=0))
        bucket.count += entry.qty
        bucket.names.append(entry.name)
        total_mv += mv * entry.qty
        counted += entry.qty

    return {
        "buckets": [buckets[mv] for mv in sorted(buckets)],
        "nonland_cards": counted,
        "average_mv": round(total_mv / counted, 2) if counted else 0.0,
    }


def color_sources(deck: Deck, cards: dict[str, CardRecord]) -> dict[str, int]:
    """How many permanents can produce each color.

    Uses Scryfall's `produced_mana`, which already accounts for lands that make
    mana through a type line (a Forest produces {G} without saying so) and for
    "add one mana of any color" effects.
    """
    counts = dict.fromkeys(COLORS, 0)
    for entry in deck.cards:
        rec = cards.get(entry.name)
        if rec is None:
            continue
        for color in getattr(rec, "produced_mana", ()) or ():
            if color in counts:
                counts[color] += entry.qty
    return counts


def commander_identity(deck: Deck, cards: dict[str, CardRecord]) -> frozenset[str]:
    """The deck's legal colors, taken from the commander's Scryfall identity.

    Never derived from mana costs -- identity already accounts for back faces,
    reminder text and land types, which a cost does not.
    """
    identity: set[str] = set()
    for name in deck.commander:
        rec = cards.get(name)
        if rec is not None:
            identity |= set(getattr(rec, "color_identity", ()) or ())
    return frozenset(identity)


def pip_requirements(deck: Deck, cards: dict[str, CardRecord]) -> list[ColorNeed]:
    """Colored pip demand against colored source supply, per color.

    This is the number that explains color screw and nothing else computes it.
    A deck can pass every identity check and still be unable to cast its own
    spells, because identity says "this card is legal here" while pips say
    "you need double black on turn three".

    Hybrid pips count toward every color that can pay them, since either will
    do. That deliberately overstates supply-side pressure rather than
    understating it.

    Only colors inside the commander's identity are reported. "Add one mana of
    any color" permanents list all five in `produced_mana`, so counting them
    naively claims a Golgari deck has ten white sources -- true of the
    permanent, meaningless for the deck, and pure noise in the chart.
    """
    pips = dict.fromkeys(COLORS, 0)
    carders = dict.fromkeys(COLORS, 0)

    for entry in deck.cards:
        rec = cards.get(entry.name)
        if entry.category == "land" or _is_land(rec):
            continue
        cost = entry.mana_cost or (getattr(rec, "mana_cost", None) if rec else None)
        seen: set[str] = set()
        for pip in parse_mana_cost(cost).pips:
            for color in pip:
                if color in pips:
                    pips[color] += entry.qty
                    seen.add(color)
        for color in seen:
            carders[color] += entry.qty

    sources = color_sources(deck, cards)
    identity = commander_identity(deck, cards)
    # Without a commander in the pool there is no identity to filter by, so
    # fall back to whatever the spells actually demand.
    relevant = identity or {c for c in COLORS if pips[c]}
    return [
        ColorNeed(color=c, pips=pips[c], sources=sources.get(c, 0), cards=carders[c])
        for c in COLORS
        if c in relevant
    ]


def category_report(deck: Deck) -> list[dict[str, Any]]:
    """Category counts against conventional targets.

    Answers "is ramp covered" with a number and a range rather than an
    opinion. Categories with no target are reported with `target: None` so the
    UI can show the count without implying a verdict.
    """
    counts = deck.category_counts
    out = []
    for category in CATEGORIES:
        count = counts.get(category, 0)
        target = CATEGORY_TARGETS.get(category)
        status = "ok"
        if target:
            low, high = target
            status = "low" if count < low else "high" if count > high else "ok"
        out.append({
            "category": category,
            "count": count,
            "target_low": target[0] if target else None,
            "target_high": target[1] if target else None,
            "status": status if target else None,
        })
    return out


def type_breakdown(deck: Deck, cards: dict[str, CardRecord]) -> dict[str, int]:
    """Counts by primary card type, taking the front face of split cards."""
    out: dict[str, int] = {}
    for entry in deck.cards:
        rec = cards.get(entry.name)
        if rec is None:
            out["Unknown"] = out.get("Unknown", 0) + entry.qty
            continue
        front = (rec.type_line or "").split(" // ")[0]
        # "Legendary Creature — Troll Warlock" -> "Creature"
        left = front.split("—")[0].strip()
        primary = left.split()[-1] if left.split() else "Unknown"
        out[primary] = out.get(primary, 0) + entry.qty
    return dict(sorted(out.items(), key=lambda kv: -kv[1]))


def deck_stats(deck: Deck, cards: dict[str, CardRecord] | None = None) -> dict[str, Any]:
    """The whole deterministic report, as plain data ready to serialise."""
    cards = cards or {}
    return {
        "slug": deck.slug,
        "name": deck.name,
        "commander": deck.commander,
        "bracket": deck.bracket,
        "total_cards": deck.total_cards,
        "land_count": deck.land_count,
        "curve": curve(deck, cards),
        "categories": category_report(deck),
        "game_changers": game_changers(deck, cards),
        "colors": [
            {
                "color": n.color,
                "pips": n.pips,
                "sources": n.sources,
                "cards": n.cards,
                "sources_per_pip": n.sources_per_pip,
            }
            for n in pip_requirements(deck, cards)
        ],
        "types": type_breakdown(deck, cards),
    }

File: go/internal/analyze/analyze.go
// Package analyze is `decks/analyze.py`: deterministic deck analysis, a pure
// function of a deck plus a card pool -- curve, pip demand against source
// supply, category coverage against conventional targets, the Game Changer
// count against the declared bracket, the opening-hand arithmetic. No
// database, no judgement that varies between runs; the same deck always
// produces the same report, and the API is a thin wrapper over it.
package analyze

import (
	"math"
	"math/big"
	"sort"
	"strconv"
	"strings"

	"github.com/aasquier/sylvan-library/go/internal/deck"
	"github.com/aasquier/sylvan-library/go/internal/mana"
	"github.com/aasquier/sylvan-library/go/internal/pool"
	"github.com/aasquier/sylvan-library/go/internal/pyfloat"
	"github.com/aasquier/sylvan-library/go/internal/reference"
)

// Bucket is one rung of the curve.
type Bucket struct {
	MV    int      `json:"mv"`
	Count int      `json:"count"`
	Names []string `json:"names"`
}

// Curve is `analyze.curve`, in the key order `service.stats_for` writes.
type Curve struct {
	AverageMV    float64  `json:"average_mv"`
	NonlandCards int      `json:"nonland_cards"`
	Buckets      []Bucket `json:"buckets"`
}

// Category is one row of `category_report`.
type Category struct {
	Category   string  `json:"category"`
	Count      int     `json:"count"`
	TargetLow  *int    `json:"target_low"`
	TargetHigh *int    `json:"target_high"`
	Status     *string `json:"status"`
}

// GameChangers is `analyze.game_changers`.
type GameChangers struct {
	Cards   []string `json:"cards"`
	Count   int      `json:"count"`
	Allowed *int     `json:"allowed"`
	Bracket *int     `json:"bracket"`
	Verdict string   `json:"verdict"`
}

// LandRow is one row of the opening hand's land distribution.
type LandRow struct {
	Lands  int     `json:"lands"`
	Chance float64 `json:"chance"`
}

// Lands is the opening hand's land odds.
type Lands struct {
	Count        int       `json:"count"`
	Distribution []LandRow `json:"distribution"`
	Keepable     float64   `json:"keepable"`
}

// CategoryOdds is one category's seen-it-by odds.
type CategoryOdds struct {
	Category      string  `json:"category"`
	Count         int     `json:"count"`
	InOpeningHand float64 `json:"in_opening_hand"`
	ByTurnFour    float64 `json:"by_turn_four"`
}

// Singleton is the chance of having seen one particular card by a turn.
type Singleton struct {
	Turn      int     `json:"turn"`
	CardsSeen int     `json:"cards_seen"`
	Chance    float64 `json:"chance"`
}

// Opening is `analyze.opening_hand`.
type Opening struct {
	DeckSize   int            `json:"deck_size"`
	HandSize   int            `json:"hand_size"`
	Lands      Lands          `json:"lands"`
	Categories []CategoryOdds `json:"categories"`
	Singleton  []Singleton    `json:"singleton"`
}

// ColorNeed is demand for one colour, and the supply available to meet it.
type ColorNeed struct {
	Color         string  `json:"color"`
	Pips          int     `json:"pips"`
	Sources       int     `json:"sources"`
	Cards         int     `json:"cards"`
	SourcesPerPip float64 `json:"sources_per_pip"`
}

// Stats is `analyze.deck_stats`, as `service.stats_for` serialises it.
type Stats struct {
	Slug         string        `json:"slug"`
	Name         string        `json:"name"`
	Commander    []string      `json:"commander"`
	Bracket      *int          `json:"bracket"`
	TotalCards   int           `json:"total_cards"`
	LandCount    int           `json:"land_count"`
	Curve        Curve         `json:"curve"`
	Categories   []Category    `json:"categories"`
	GameChangers GameChangers  `json:"game_changers"`
	Opening      Opening       `json:"opening"`
	Colors       []ColorNeed   `json:"colors"`
	Types        orderedCounts `json:"types"`
}

// orderedCounts serialises as an object in its own order -- Python's dict
// keeps the order `type_breakdown` sorted it into, and a Go map would not.
type orderedCounts struct {
	Keys   []string
	Counts map[string]int
}

// MarshalJSON writes the counts as an object, highest first.
func (o orderedCounts) MarshalJSON() ([]byte, error) {
	var b strings.Builder
	b.WriteByte('{')
	for i, k := range o.Keys {
		if i > 0 {
			b.WriteByte(',')
		}
		b.WriteString(strconv.Quote(k))
		b.WriteByte(':')
		b.WriteString(strconv.Itoa(o.Counts[k]))
	}
	b.WriteByte('}')
	return []byte(b.String()), nil
}

// isLand is `analyze._is_land`: the whole type line, as Python reads it here.
func isLand(rec *pool.CardRecord) bool {
	return rec != nil && strings.Contains(rec.TypeLine, "Land")
}

// round2 is Python's `round(x, 2)`: half to even on the scaled value.
func round2(x float64) float64 {
	return math.RoundToEven(x*100) / 100
}

func costOf(entry deck.CardEntry, rec *pool.CardRecord) string {
	if entry.ManaCost != nil && *entry.ManaCost != "" {
		return *entry.ManaCost
	}
	if rec != nil && rec.ManaCost != nil {
		return *rec.ManaCost
	}
	return ""
}

// CurveOf is `analyze.curve`: the MV histogram over the nonland 99.
func CurveOf(d *deck.Deck, cards map[string]*pool.CardRecord) Curve {
	buckets := map[int]*Bucket{}
	totalMV, counted := 0, 0
	for _, entry := range d.Cards {
		rec := cards[entry.Name]
		if entry.Category == "land" || isLand(rec) {
			continue
		}
		mv := mana.Parse(costOf(entry, rec)).ManaValue()
		b, ok := buckets[mv]
		if !ok {
			b = &Bucket{MV: mv, Names: []string{}}
			buckets[mv] = b
		}
		b.Count += entry.Qty
		b.Names = append(b.Names, entry.Name)
		totalMV += mv * entry.Qty
		counted += entry.Qty
	}
	mvs := make([]int, 0, len(buckets))
	for mv := range buckets {
		mvs = append(mvs, mv)
	}
	sort.Ints(mvs)
	out := Curve{NonlandCards: counted, Buckets: []Bucket{}}
	for _, mv := range mvs {
		out.Buckets = append(out.Buckets, *buckets[mv])
	}
	if counted > 0 {
		out.AverageMV = round2(float64(totalMV) / float64(counted))
	}
	return out
}

// ColorSources is `analyze.color_sources`: how many permanents can produce
// each colour, from Scryfall's `produced_mana`.
func ColorSources(d *deck.Deck, cards map[string]*pool.CardRecord) map[string]int {
	counts := map[string]int{}
	for _, c := range mana.Colors {
		counts[c] = 0
	}
	for _, entry := range d.Cards {
		rec := cards[entry.Name]
		if rec == nil {
			continue
		}
		for _, color := range rec.ProducedMana {
			if _, ok := counts[color]; ok {
				counts[color] += entry.Qty
			}
		}
	}
	return counts
}

// CommanderIdentity is `analyze.commander_identity`: the deck's legal
// colours from the commander's Scryfall identity, never from mana costs.
func CommanderIdentity(d *deck.Deck, cards map[string]*pool.CardRecord) map[string]bool {
	identity := map[string]bool{}
	for _, name := range d.Commander {
		if rec := cards[name]; rec != nil {
			for _, c := range rec.ColorIdentity {
				identity[c] = true
			}
		}
	}
	return identity
}

// PipRequirements is `analyze.pip_requirements`: coloured pip demand against
// coloured source supply, per colour inside the commander's identity.
func PipRequirements(d *deck.Deck, cards map[string]*pool.CardRecord) []ColorNeed {
	pips := map[string]int{}
	carders := map[string]int{}
	for _, entry := range d.Cards {
		rec := cards[entry.Name]
		if entry.Category == "land" || isLand(rec) {
			continue
		}
		seen := map[string]bool{}
		for _, pip := range mana.Parse(costOf(entry, rec)).Pips {
			for _, color := range pip {
				if isColor(color) {
					pips[color] += entry.Qty
					seen[color] = true
				}
			}
		}
		for color := range seen {
			carders[color] += entry.Qty
		}
	}
	sources := ColorSources(d, cards)
	identity := CommanderIdentity(d, cards)
	relevant := identity
	if len(relevant) == 0 {
		relevant = map[string]bool{}
		for _, c := range mana.Colors {
			if pips[c] > 0 {
				relevant[c] = true
			}
		}
	}
	out := []ColorNeed{}
	for _, c := range mana.Colors {
		if !relevant[c] {
			continue
		}
		need := ColorNeed{Color: c, Pips: pips[c], Sources: sources[c], Cards: carders[c]}
		if need.Pips > 0 {
			need.SourcesPerPip = round2(float64(need.Sources) / float64(need.Pips))
		}
		out = append(out, need)
	}
	return out
}

func isColor(c string) bool {
	for _, col := range mana.Colors {
		if col == c {
			return true
		}
	}
	return false
}

// CategoryReport is `analyze.category_report`: counts against conventional
// targets, reported with `target: None` where there is no target.
func CategoryReport(d *deck.Deck) []Category {
	counts, _ := d.CategoryCounts()
	model := reference.Deck()
	out := []Category{}
	for _, category := range model.Categories {
		row := Category{Category: category, Count: counts[category]}
		if target, ok := model.CategoryTargets[category]; ok && len(target) == 2 {
			lo, hi := target[0], target[1]
			row.TargetLow, row.TargetHigh = &lo, &hi
			status := "ok"
			if row.Count < lo {
				status = "low"
			} else if row.Count > hi {
				status = "high"
			}
			row.Status = &status
		}
		out = append(out, row)
	}
	return out
}

// TypeBreakdown is `analyze.type_breakdown`: counts by primary card type,
// taking the front face, highest first (ties in first-seen order).
func TypeBreakdown(d *deck.Deck, cards map[string]*pool.CardRecord) orderedCounts {
	counts := map[string]int{}
	order := []string{}
	bump := func(key string, n int) {
		if _, seen := counts[key]; !seen {
			order = append(order, key)
		}
		counts[key] += n
	}
	for _, entry := range d.Cards {
		rec := cards[entry.Name]
		if rec == nil {
			bump("Unknown", entry.Qty)
			continue
		}
		front, _, _ := strings.Cut(rec.TypeLine, " // ")
		left, _, _ := strings.Cut(front, "—")
		fields := strings.Fields(strings.TrimSpace(left))
		primary := "Unknown"
		if len(fields) > 0 {
			primary = fields[len(fields)-1]
		}
		bump(primary, entry.Qty)
	}
	sort.SliceStable(order, func(i, j int) bool { return counts[order[i]] > counts[order[j]] })
	return orderedCounts{Keys: order, Counts: counts}
}

// GameChangersOf is `analyze.game_changers`: which Game Changers the deck
// runs and what its bracket allows. `unknown` when the deck declares no
// bracket or the pool is missing -- an absent count is not a count of zero.
func GameChangersOf(d *deck.Deck, cards map[string]*pool.CardRecord) GameChangers {
	names := []string{}
	for _, c := range d.Cards {
		names = append(names, c.Name)
	}
	names = append(names, d.Commander...)
	if d.Companion != nil {
		names = append(names, *d.Companion)
	}
	found := []string{}
	for _, n := range names {
		if rec := cards[n]; rec != nil && rec.GameChanger {
			found = append(found, n)
		}
	}
	sort.Strings(found)
	out := GameChangers{Cards: found, Count: len(found), Bracket: d.Bracket}
	bracketKey := "0"
	if d.Bracket != nil {
		bracketKey = strconv.Itoa(*d.Bracket)
	}
	allowed, known := reference.Deck().GameChangerLimits[bracketKey]
	if known {
		out.Allowed = allowed
	}
	switch {
	case len(cards) == 0 || d.Bracket == nil:
		out.Verdict = "unknown"
	case out.Allowed == nil || len(found) <= *out.Allowed:
		out.Verdict = "ok"
	default:
		out.Verdict = "over"
	}
	return out
}

// comb is math.comb: 0 for k > n or k < 0.
func comb(n, k int) *big.Int {
	if k < 0 || k > n {
		return big.NewInt(0)
	}
	return new(big.Int).Binomial(int64(n), int64(k))
}

func ratio(a, b *big.Int) float64 {
	if b.Sign() == 0 {
		return 0
	}
	f, _ := new(big.Rat).SetFrac(a, b).Float64()
	return f
}

// atLeastOne is `analyze._at_least_one`: P(at least one of `copies` among
// `seen` draws from `deckSize`), hypergeometric complement.
func atLeastOne(deckSize, copies, seen int) float64 {
	if copies <= 0 || deckSize <= 0 {
		return 0
	}
	if seen > deckSize {
		seen = deckSize
	}
	total := comb(deckSize, seen)
	if total.Sign() == 0 {
		return 0
	}
	return 1 - ratio(comb(deckSize-copies, seen), total)
}

// OpeningHand is `analyze.opening_hand`: draw odds for the opening seven and
// the seen-it-by-turn table, on the draw (`7 + t` cards by the end of turn t).
//
// `keepable` goes through `pyfloat.Fsum` and not a `+=` loop, matching the
// `math.fsum` Python spells it with. The loop was the obvious transcription of
// Python's old `sum(...)` and it was wrong in a way that only fires on one of
// the two interpreters: `sum()` over floats is compensated from CPython 3.12
// and left to right before it, so a Go accumulation loop reproduces **3.11**
// while the container runs 3.12. Three terms is enough to see it -- swept over
// every deck size from 8 to 250, the two arithmetics disagree in 5,098 shapes,
// 33 of them ordinary Commander decks, and the difference reaches the JSON.
func OpeningHand(d *deck.Deck) Opening {
	n := d.TotalCards()
	hand := 7
	if n < hand {
		hand = n
	}
	lands := d.LandCount()
	total := comb(n, hand)
	dist := []LandRow{}
	keep := []float64{}
	for k := 0; k <= hand; k++ {
		chance := 0.0
		if total.Sign() != 0 {
			num := new(big.Int).Mul(comb(lands, k), comb(n-lands, hand-k))
			chance = ratio(num, total)
		}
		dist = append(dist, LandRow{Lands: k, Chance: chance})
		if k >= 2 && k <= 4 {
			keep = append(keep, chance)
		}
	}
	keepable := pyfloat.Fsum(keep)
	counts, order := d.CategoryCounts()
	sort.SliceStable(order, func(i, j int) bool { return counts[order[i]] > counts[order[j]] })
	categories := []CategoryOdds{}
	for _, cat := range order {
		m := counts[cat]
		categories = append(categories, CategoryOdds{Category: cat, Count: m,
			InOpeningHand: atLeastOne(n, m, hand), ByTurnFour: atLeastOne(n, m, hand+4)})
	}
	singleton := []Singleton{}
	for _, t := range []int{1, 4, 7, 10} {
		seen := hand + t
		if seen > n {
			seen = n
		}
		chance := 0.0
		if n > 0 {
			chance = float64(seen) / float64(n)
		}
		singleton = append(singleton, Singleton{Turn: t, CardsSeen: seen, Chance: chance})
	}
	return Opening{DeckSize: n, HandSize: hand,
		Lands:      Lands{Count: lands, Distribution: dist, Keepable: keepable},
		Categories: categories, Singleton: singleton}
}

// DeckStats is `analyze.deck_stats`: the whole deterministic report.
func DeckStats(d *deck.Deck, cards map[string]*pool.CardRecord) Stats {
	if cards == nil {
		cards = map[string]*pool.CardRecord{}
	}
	return Stats{
		Slug: d.Slug, Name: d.Name, Commander: append([]string{}, d.Commander...), Bracket: d.Bracket,
		TotalCards: d.TotalCards(), LandCount: d.LandCount(),
		Curve: CurveOf(d, cards), Categories: CategoryReport(d), GameChangers: GameChangersOf(d, cards),
		Opening: OpeningHand(d), Colors: PipRequirements(d, cards), Types: TypeBreakdown(d, cards),
	}
}
